bmt/bmt.go

```diff
@@ -0,0 +1,571 @@
+	count       int         // segment count
+	size        int         // for hash.Hash same as hashsize
+	cur         int         // cursor position for righmost currently open chunk
+	segment     []byte      // the rightmost open segment (not complete)
+	depth       int         // index of last level
+	result      chan []byte // result channel
+	hash        []byte      // to record the result
+	max         int32       // max segments for SegmentWriter interface
+	blockLength []byte      // The block length that needes to be added in Sum
+}
+
+// creates a reusable BMTHasher
+// implements the hash.Hash interface
+// pulls a new BMTree from a resource pool for hashing each chunk
+func NewBMTHasher(p *BMTreePool) *BMTHasher {
+	return &BMTHasher{
+		pool:      p,
+		depth:     depth(p.SegmentCount),
+		size:      p.SegmentSize,
+		blocksize: p.SegmentSize,
+		count:     p.SegmentCount,
+		result:    make(chan []byte),
+	}
+}
+
+// a reuseable segment hasher representing a node in a BMT
+// it allows for continued writes after a Sum
+// and is left in completely reusable state after Reset
+type BMTNode struct {
+	level, index int      // position of node for information/logging only
+	initial      bool     // first and last node
+	root         bool     // whether the node is root to a smaller BMT
+	isLeft       bool     // whether it is left side of the parent double segment
+	unbalanced   bool     // indicates if a node has only the left segment
+	parent       *BMTNode // BMT connections
+	state        int32    // atomic increment impl concurrent boolean toggle
+	left, right  []byte
+}
+
+// constructor for segment hasher nodes in the BMT
+func NewBMTNode(level, index int, parent *BMTNode) *BMTNode {
+	self := &BMTNode{
+		parent:  parent,
+		level:   level,
+		index:   index,
+		initial: index == 0,
+		isLeft:  index%2 == 0,
+	}
+	return self
+}
+
+// provides a pool of BMTrees used as resources by BMTHasher
+// a BMTree popped from the pool is guaranteed to have clean state
+// for hashing a new chunk
+// BMTHasher Reset releases the BMTree to the pool
+type BMTreePool struct {
+	lock         sync.Mutex
+	c            chan *BMTree
+	hasher       Hasher
+	SegmentSize  int
+	SegmentCount int
+	Capacity     int
+	count        int
+}
+
+// create a BMTree pool with hasher, segment size, segment count and capacity
+// on GetBMTree it reuses free BMTrees or creates a new one if size is not reached
+func NewBMTreePool(hasher Hasher, segmentCount, capacity int) *BMTreePool {
+	segmentSize := hasher().Size()
+	return &BMTreePool{
+		c:            make(chan *BMTree, capacity),
+		hasher:       hasher,
+		SegmentSize:  segmentSize,
+		SegmentCount: segmentCount,
+		Capacity:     capacity,
+	}
+}
+
+// drains the pool uptil it has no more than n resources
+func (self *BMTreePool) Drain(n int) {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+	for len(self.c) > n {
+		select {
+		case <-self.c:
+			self.count--
+		default:
```

Comment:
This `default` clause is redundant, it will just lead to the next loop iteration trying to select on the channel again, so it may as well just be:

```go
for len(self.c) > n {
	<-self.c
	self.count--
}
```